### DISCUSSION TITLE
Controlling faction reputation?

### DISCUSSION THREAD

**Discussion by CMDR-Evolution** [EDCD/EDDI#2254] [Needs thought]
Hello EDDI,

I absolutely love your program. It's totally essential for the full Elite experience and your work is astounding. I'm not very good at coding but over the years, I've made loads of adaptions to commands for more info, variance etc.

I'm currently overhauling my personalised EDDI profile to make sure it's got everything covered by updates and noticed there's a blank command for reputation. (Sorry, I don't know the exact name as I'm using my phone to type this).

What I'm wondering, is if there's a way to hear my reputation when I'm travelling to systems please? If so, I'll then create an IF condition to limit it to anything not neutral.

As I've played Elite for around 6,400 hours, I'd get a lot of enjoyment from hearing all the places I've had an impact over the years!

Thank you so much (even if this can't be done!)

**Comment by Tkael** [comment]
EDDI should be able to provide that. You'd probably want to write something like:

```
{for faction in system.factions:
    {if faction.myreputation > 50:
        {_ Do something }
    }
}
```

Reputation ranges from -100 (fully hostile) to +100 (fully allied). I'm not sure about the precise percentages that correspond to cordial, friendly, etc. but I'm sure those thresholds can be worked out easily enough.

Please also note that reputation is sourced exclusively from the journal - player-specific data like a commander's reputation with various factions is not shared over EDDN or available from EDSM. Consequently, you'll only be able to retrieve reputation data for systems that you've recently visited (for example, after `Jumped`, `Carrier jumped` and/or `Location` events).

If you could post your final script once you find something that you're happy with, I'm sure other commanders will be thankful. :-)

**Comment by CMDR-Evolution** [comment]
Oh wow, thank you Tkael,

This isn't the first time you've been a legend from one of my questions.... there's been a few varied ones over the years and your speedy insightful help is always appreciated! You really are a coding master and I wish I had even a small amount of your genius skills! 

That's a massive help. It may take a while to get all the exact boundary values but when I know for sure, I'll be sure to drop an update here with the results.

Thanks again :)

**Reply by CMDR-Evolution** [reply]
Hello again Tkael,

I've got it sorted! I've added it to the system report and location commands, which I'll share below.

You'll notice there's also some small other changes to the system command, to avoid it saying a system is e.g. an independent communism/corporate/etc but now says society after those words, followed by the new reputation command after the controlling faction,

- Initially, the reputation part was making the voice repeat the output twice.  Further testing found that the rep gets called 7 times in the EDDI mock data. Placing it inside a function and calling the function in a 2nd section makes it work once, as expected.
- I've done it so it doesn't say anything for neutral - so only ones where there's been an impact with the controlling faction are mentioned. 
- The command says the reputation -30% slower as it didn't sound quite right at full speed.
- Note: I've had to use slightly different boundary values as of course with any boundary partitioning, there shouldn't be the same number within two categories - like Elite uses for rep. I assume this is why we get messages in-game saying we're allied etc when we're still actually friendly.
- Note: If you view the output in the EDDI log file, you'll notice there's a space after rep/reputation. I tried removing that and having the space on only the OneOf 'is' which follows, but that doesn't work.  Although, it is spoken correctly so a non-issue IMO, 

**Here's the adjusted system report command:**
```
{_ Fetch from Context, if a context is set _}
{if state.eddi_context_system_name:
    {set reportSystem to SystemDetails(state.eddi_context_system_name)}
|else:
    {set reportSystem to SystemDetails(system.systemname)}
}

{if !reportSystem.name || reportbody.name = "":
    I'm not sure which system you are asking about.
|else:
    {P(reportSystem.name)} is

    {if reportSystem.population && reportSystem.population > 0:
        {if reportSystem.security = "High":
            {OneOf("a highly policed", "a well regulated")}
        |elif reportSystem.security = "Low":
            {OneOf("a poorly policed", "a poorly secured")}
        |elif reportSystem.security = "Anarchy":
            {OneOf("an unpoliced", "an unsecured")}
        |else:
            {if reportSystem.allegiance = "Federation": a |else: an }
        }
        {if reportSystem.allegiance = "Federation":
            Federal
        |elif reportSystem.allegiance = "Empire":
            Imperial
        |elif reportSystem.allegiance = "Alliance":
            Alliance
        |else:
            independent
        }
        {if reportSystem.government = "Communism": Communist society
        |elif reportSystem.government = "Cooperative": Cooperative society
		|elif reportSystem.government = "Corporate": Corporate society 
		|elif reportSystem.government = "Feudal": Feudal society
		|elif reportSystem.government = "Imperial": Imperial society
		|elif reportSystem.government = "Patronage": Patronage society
		|elif reportSystem.government = "None": system
		|else: {reportSystem.government}
        },
        
        {OneOf("of", "with a population of", "home to", "supporting")} {Humanise(reportSystem.population)} {OneOf("people", "souls","residents","citizens")}
        {if reportSystem.powerState.invariantName != "None":
            {if reportSystem.powerState.invariantName = "Home System":
                , and is home to {P(reportSystem.power, "power")}
            |elif reportSystem.powerState.invariantName = "Control":
                , and is {OneOf("allied with", "controlled by")} {P(reportSystem.power, "power")}
            |elif reportSystem.powerState.invariantName = "Exploited":
                , and is {OneOf("aligned to", "within the influence of")} {P(reportSystem.power, "power")}
            |elif reportSystem.powerState.invariantName = "Contested":
                , and is contested by multiple powers
            }
        }.
        {if reportSystem.faction:
            {P(reportSystem.faction, "faction")} is the {OneOf("controlling", "local", "immediate")} {OneOf("power", "faction")},
        }
		{for faction in system.factions:
			{if faction.myreputation >= 90 && faction.myreputation <= 100: 
				{set rep to SpeechRate("allied", "-30%")}
			|elif faction.myreputation >= 35 && faction.myreputation <= 89: 
				{set rep to SpeechRate("friendly", "-30%")}
			|elif faction.myreputation >= 4 && faction.myreputation <= 34: 
				{set rep to SpeechRate("cordial", "-30%")}
			|elif faction.myreputation >= -89 && faction.myreputation <= -35: 
				{set rep to SpeechRate("unfriendly", "-30%")}
			|elif faction.myreputation >= -100 && faction.myreputation <= -90: 
				{set rep to SpeechRate("hostile", "-30%")}
			}
		}
		{if reportSystem.faction != "None":		
			{if faction.myreputation >= 4:
				with whom your {OneOf("rep", "reputation")} {OneOf("'s"," is")} {rep}.
			|elif faction.myreputation <= -35:
				with whom your  {OneOf("rep", "reputation")} {OneOf("'s"," is")} {rep}.
			|elif faction.myreputation <= -90:
				with whom your  {OneOf("rep", "reputation")} {OneOf("'s"," is")} {rep}
				, so be careful here {F('Honorific')}.
			}
        }

        {if len(reportSystem.stations) = 0:
            There are no stations present
        |else:
            {set systemStations to []}
            {if len(reportSystem.orbitalstations) = 1:
                {set orbitalStationDesc to:                
                    There is {OneOf("a single", "a lone", "just one", "one")} orbital station
                }
            |elif len(reportSystem.orbitalstations) > 1:
                {set orbitalStationDesc to:                
                    There are {len(reportSystem.orbitalstations)} orbital stations
                }
            }
            {if len(reportSystem.planetarystations) = 1:
                {set planetaryStationDesc to:            
                    {if len(reportSystem.orbitalstations) = 0: There is}
                    {OneOf("a single", "a lone", "just one", "one")} planetary station
                }
            |elif len(reportSystem.planetarystations) > 1:
                {set planetaryStationDesc to:            
                    {if len(reportSystem.orbitalstations) = 0: There are}
                    {len(reportSystem.planetarystations)} planetary stations
                }
            }
            {if len(reportSystem.carriersignalsources) = 1:
                {set fleetCarrierDesc to:
                    {if len(reportSystem.orbitalstations) = 0 && len(reportSystem.planetarystations) = 0: There is}
                    {OneOf("a single", "a lone", "just one", "one")} fleet carrier
                }
            |elif len(reportSystem.carriersignalsources) > 1:
                {set fleetCarrierDesc to:            
                    {if len(reportSystem.orbitalstations) = 0 && len(reportSystem.planetarystations) = 0: There are}
                    {len(reportSystem.carriersignalsources)} fleet carriers
                }
            }
            {if orbitalStationDesc:
                {set systemStations to cat(systemStations, [orbitalStationDesc])}
            }
            {if planetaryStationDesc:
                {set systemStations to cat(systemStations, [planetaryStationDesc])}
            }
            {if fleetCarrierDesc:
                {set systemStations to cat(systemStations, [fleetCarrierDesc])}
            }
            {List(systemStations)}
        }
        in this system
    |else:
       unpopulated
   }.
}
```


**Then here is the adjusted location command as well:**
```
{_ Context _}
{SetState('eddi_context_last_subject', 'location')}
{SetState('eddi_context_system_name', event.system)}
{SetState('eddi_context_body_name', event.body)}
{SetState('eddi_context_body_type', event.bodytype)}
{SetState('eddi_context_body_system', event.system)}


{OneOf("You are","You're")}
{if event.docked || event.station = event.body:
    {set station to StationDetails(event.station, event.system)}
    {set station_missions to F("Mission check station")}
    {set station_crimes to F("Crime check station")}

    {if event.docked:
        docked at
    |else:
        {OneOf("near", "close to", "in the vicinity of")}
    }
    {P(station.model, "station")}, {P(event.station, "station")},
    {if event.station != event.body:
        on {P(event.body, "body")},
    }
|elif event.body:
    {set system_missions to F("Mission check system")}
    {set system_crimes to F("Crime check system")}
    {OneOf("near", "close to", "in the vicinity of")} {P(event.body, "body")},
}
in the {P(event.system, "starsystem")} system,

		{for faction in system.factions:
			{if faction.myreputation >= 90 && faction.myreputation <= 100: 
				{set rep to SpeechRate("allied", "-30%")}
			|elif faction.myreputation >= 35 && faction.myreputation <= 89: 
				{set rep to SpeechRate("friendly", "-30%")}
			|elif faction.myreputation >= 4 && faction.myreputation <= 34: 
				{set rep to SpeechRate("cordial", "-30%")}
			|elif faction.myreputation >= -89 && faction.myreputation <= -35: 
				{set rep to SpeechRate("unfriendly", "-30%")}
			|elif faction.myreputation >= -100 && faction.myreputation <= -90: 
				{set rep to SpeechRate("hostile", "-30%")}
			}
		}
		{if reportSystem.faction != "None":		
			{if faction.myreputation >= 4:
			where your {OneOf("rep", "reputation")} {OneOf("'s","is")} {rep}.
			|elif faction.myreputation <= -35:
			where your {OneOf("rep", "reputation")} {OneOf("'s","is")} {rep}.
			|elif faction.myreputation <= -90:
			where your {OneOf("rep", "reputation")} {OneOf("'s","is")} {rep}
			, so be careful here {F('Honorific')}.
			}
        }

{if station_missions || station_crimes ||system_missions || system_crimes:
    {Pause(500)}
    Information:
    {if station_missions:
        {Pause(500)}
        {station_missions}
    }
    {if station_crimes:
        {Pause(500)}
        {station_crimes}
    }
    {if system_missions:
        {Pause(500)}
        {system_missions}
    }
    {if system_crimes:
        {Pause(500)}
        {system_crimes}
    }
}
```

**Comment by CMDR-Evolution** [comment]
Ah, so no, it's not working properly.

For some unknown reason, it still says the reputation part for neutral controlling factions although the 2nd IF statement (for speech) should exclude that?
Of course, the 1st IF statement doesn't have an elif for neutral, so the speech output just says your reputation is __. 

Considering the reputation range is - 100 to 100, further testing found that the second IF statement could be set even be greater than 400 and yet, it still triggers the command. As you can see in the code, they're running from the same event so should work the same way AFAIK. I'm so confused. 

I have no idea and can't see any mistakes (from my lack of any dev experience), sorry. 

Any help would be massively appreciated and thank you if you can shed any light

**Reply by CMDR-Evolution** [reply]
Like the last 1.5 weeks, I spent a lot of time investigating and may have found solutions without residual issues cropping up. Preliminary testing has passed but more test conditions are required.

As I've been doing for 5 days now, once I've thoroughly tested it and if I haven't heard back, I'll remove and edit comments to reflect the current situation :)

P.S. I hope the level of detail has been helpful? I'm a Senior Manual QA Engineer so I always try to explain all angles ;).

Cheers

**Reply by Tkael** [reply]
Looking over the snippet
```
		{for faction in system.factions:
			{if faction.myreputation >= 90 && faction.myreputation <= 100: 
				{set rep to SpeechRate("allied", "-30%")}
			|elif faction.myreputation >= 35 && faction.myreputation <= 89: 
				{set rep to SpeechRate("friendly", "-30%")}
			|elif faction.myreputation >= 4 && faction.myreputation <= 34: 
				{set rep to SpeechRate("cordial", "-30%")}
			|elif faction.myreputation >= -89 && faction.myreputation <= -35: 
				{set rep to SpeechRate("unfriendly", "-30%")}
			|elif faction.myreputation >= -100 && faction.myreputation <= -90: 
				{set rep to SpeechRate("hostile", "-30%")}
			}
		}
		{if reportSystem.faction != "None":		
			{if faction.myreputation >= 4:
				with whom your {OneOf("rep", "reputation")} {OneOf("'s"," is")} {rep}.
			|elif faction.myreputation <= -35:
				with whom your  {OneOf("rep", "reputation")} {OneOf("'s"," is")} {rep}.
			|elif faction.myreputation <= -90:
				with whom your  {OneOf("rep", "reputation")} {OneOf("'s"," is")} {rep}
				, so be careful here {F('Honorific')}.
			}
        }
```

You are looping through your rep with each faction and setting a single `rep` value, overwriting the prior value each time. That means that for each faction you are overwriting the value from the previous iteration of the loop. The rep with the last faction in the loop (regardless of whether that is the controlling faction) will be reported.

Proposed change for the `System report` (which I shall also test):
```
        {if reportSystem.faction:
            {for faction in reportSystem.factions:
                {if reportSystem.faction = faction.name:
			        {if faction.myreputation >= 75: 
			        	{set rep to "allied")}
			        |elif faction.myreputation >= 35: 
				        {set rep to "friendly")}
			        |elif faction.myreputation >= 15: 
				        {set rep to "cordial")}
			        |elif faction.myreputation >= -15: 
			            {set rep to ""} {_ Neutral }
			        |elif faction.myreputation >= -90:
			            {set rep to "unfriendly"}
			        |elif faction.myreputation >= -100: 
				        {set rep to "hostile"}
			        }
                }
		    }
		    {if len(rep) > 0:
			    You are {SpeechRate(rep, "-30%")} with the {OneOf("controlling", "local")} {OneOf("power", "faction")}, {P(reportSystem.faction, "faction")}. 
			    {if rep = "hostile": Be careful here {F('Honorific')}.}
			|else:
                {P(reportSystem.faction, "faction")} is the {OneOf("controlling", "local")} {OneOf("power", "faction")}.	
			}
        }
```
(positive rep levels are taken from https://elite-dangerous.fandom.com/wiki/Reputation, negative rep levels are speculative)

**Reply by CMDR-Evolution** [reply]
Thank you so much @Tkael ,

I'd made some changes on my side and was going to update the thread later. They were only small and I hadn't realised I didn't constructed it correctly, so I really appreciate your insightful help. I'll start using and testing your much better script instead when I finish work today o7. 

One thing I've noticed is that when we travel to new systems, my script was saying "with whom your rep is ___", but then once the system is reached, then the correct {rep} output would work. 
I presume this goes back to your initial advice that the rep only works once the journal file is updated? 
I assume that means, it can't be used in the system report script, but only the location one?

Also, I have an unrelated question so rather than adding it here, I've raised it separately here: 
[Planet for Surface signals detected #2258](https://github.com/EDCD/EDDI/discussions/2258)

Thank you again

**Reply by Tkael** [reply]
It would need to trigger after the `Jumped` event to pick up faction reputations. I've been noticing a quicker load time when jumping between systems recently so I believe I've already reconfigured my personal personality to invoke `System report` after the jump rather than during.

**Reply by Tkael** [reply]
I'll also give some thought to the other question you raised (thanks for keeping questions separated).

**Reply by CMDR-Evolution** [reply]
Thank you Tkael. I really appreciate it :)

P.S I've taken the old script out of the system report script and added it into 'jumped', just after it says Information 👍

**Reply by CMDR-Evolution** [reply]
Hello again Tkael,

I hope you're doing well.

I was hoping to find a solution myself but I'm unfortunately stuck. For some reason, the 'jumped' script works perfectly but the 'location' script is seeming to never work now. I'm not sure if something changed on the Frontier side, perhaps. Anyway, even directly referencing the command isn't doing anything now - and I guess, seeing as it all stems from that, that could be why it's not working:
{reportSystem.faction} 
I also tried the other commands, tried removing report etc. 

Lastly and separately, when I've got EDDI sorted, I was considering sharing my personality and wondered if here's the best place? I searched for a long time online but couldn't seem to find any personalities, although I'm sure I'd seen people sharing them in the past. Would the github personality section be the best place for that? https://github.com/EDCD/EDDI/discussions/categories/share-personalities

**Reply by Tkael** [reply]
Guessing here but perhaps you haven't used `{SystemDetails()}` to define a `reportSystem` object in your `Location` event? You can use the `system` object instead as an alternative.

```
{if system.faction:
    {for faction in system.factions:
        {if system.faction = faction.name:
		    {if faction.myreputation >= 75: 
		        {set rep to "allied")}
			|elif faction.myreputation >= 35: 
                {set rep to "friendly")}
			|elif faction.myreputation >= 5: 
                {set rep to "cordial")}
			|elif faction.myreputation >= -15: 
                {set rep to ""} {_ Neutral }
			|elif faction.myreputation >= -90:
                {set rep to "unfriendly"}
			|elif faction.myreputation >= -100: 
                {set rep to "hostile"}
			}
        }
    }
    {if len(rep) > 0:
        You are {rep} with the {OneOf("controlling", "local")} {OneOf("power", "faction")}, {P(system.faction, "faction")}. 
        {if rep = "hostile": Be careful here {F('Honorific')}.}
    |else:
        {P(system.faction, "faction")} is the {OneOf("controlling", "local")} {OneOf("power", "faction")}.	
    }
}
```

Traditionally people have shared scripts and personalities on the forums (<https://forums.frontier.co.uk/threads/eddi-scripts-and-eddi-enabled-va-commands-thread.311614/page-70#post-9535642>) but I'd like to encourage posting scripts here in Github where they may be a little easier to find.

**Reply by CMDR-Evolution** [reply]
Yet again, you've been a lifesaver and your advice works perfectly, thank you.

Regarding the personalities, yeah, I had a look at that forum thread but found it difficult to filter out non VA related info and EDDI personalities there. I'll upload it to github so it's easier to find and in case anyone finds it useful o7
